sentry-ruby/spec/sentry/session_flusher_spec.rb

```diff
@@ -0,0 +1,99 @@
+require "spec_helper"
+
+RSpec.describe Sentry::SessionFlusher do
+  let(:string_io) { StringIO.new }
+
+  let(:configuration) do
+    Sentry::Configuration.new.tap do |config|
+      config.release = 'test-release'
+      config.environment = 'test'
+      config.transport.transport_class = Sentry::DummyTransport
+      config.background_worker_threads = 0
+      config.logger = Logger.new(string_io)
+    end
+  end
+
+  let(:client) { Sentry::Client.new(configuration) }
+  let(:transport) { client.transport }
+  subject { described_class.new(configuration, client) }
+
+  before do
+    Sentry.background_worker = Sentry::BackgroundWorker.new(configuration)
+  end
+
+  describe "#initialize" do
+    context "when config.release is nil" do
+      before { configuration.release = nil }
+
+      it "logs debug message" do
+        flusher = described_class.new(configuration, client)
+
+        expect(string_io.string).to match(
+          /Sessions won't be captured without a valid release/
+        )
+      end
+    end
+  end
+
+  describe "#flush" do
+    it "early returns with no pending_aggregates" do
+      subject.instance_variable_set(:@pending_aggregates, {})
+
+      expect do
+        subject.flush
+      end.not_to change { transport.envelopes.count }
+    end
+
+    it "captures pending_aggregates in background worker" do
+      t1 = Sentry.utc_now
+      t1_bucket = Time.utc(t1.year, t1.month, t1.day, t1.hour, t1.min)
+
+      aggregates = { t1_bucket => { exited: 50, errored: 10 } }
+      subject.instance_variable_set(:@pending_aggregates, aggregates)
```

Comment:
Again I'm not sure how I should rspec properly, so..
There are already higher level integration tests in `capture_exceptions_spec`. This block is for testing the behaviour of `#flush`. If I initiate `Session` objects and add them via `add_session` in a `before` block, I'm testing more that just the `#flush` method, so not sure what the right way to approach this is.

Reply:
I think it's not related to the test framework and it's more like a personal opinion tbh.
Here are 2 options:
1. Set up tests with public APIs, which will involve other parts into the tests.
2. Set up tests with private APIs, which will bind the test with implementation details.

Generally I prefer 1) because when we refactor implementations, 2) usually requires us to change the tests as well. And that'll increase the chance of breaking things without knowing it.

The downside of 1) is that when the public API we use for test setup failed (`#add_session`), this test will break as well. The worst case of this is that we won't be sure about whether the cause is on `add_session` malfunction or it's caused by `flush`. But given that we also test `add_session`, it should be easy to distinguish.

Reply:
done